x-pack/plugin/security/qa/jwt-realm/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRestIT.java

```diff
@@ -0,0 +1,458 @@
+        return signHmacJwt(claimsSet);
+    }
+
+    private JWTClaimsSet buildJwtForRealm2(String principal, Instant issueTime) {
+        final String emailAddress = principal + "@" + randomAlphaOfLengthBetween(3, 6) + ".example.com";
+        // The "jwt2" realm, supports 3 audiences (es01/02/03)
+        final String audience = "es0" + randomIntBetween(1, 3);
+        final JWTClaimsSet claimsSet = buildJwt(
+            Map.ofEntries(Map.entry("iss", "my-issuer"), Map.entry("aud", audience), Map.entry("email", emailAddress)),
+            issueTime
+        );
+        return claimsSet;
+    }
+
+    private SignedJWT signRsaJwt(JWTClaimsSet claimsSet) throws IOException, JOSEException, ParseException {
+        final RSASSASigner signer = loadRsaSigner();
+        return signJWt(signer, "RS256", claimsSet);
+    }
+
+    private RSASSASigner loadRsaSigner() throws IOException, ParseException, JOSEException {
+        // The "jwt1" realm is configured using public JWKSet (in build.gradle)
+        try (var in = getDataInputStream("/jwk/rsa-private-jwkset.json")) {
+            final JWKSet jwkSet = JWKSet.load(in);
+            final JWK key = jwkSet.getKeyByKeyId("test-rsa-key");
+            assertThat(key, instanceOf(RSAKey.class));
+            return new RSASSASigner((RSAKey) key);
+        }
+    }
+
+    private SignedJWT signHmacJwt(JWTClaimsSet claimsSet) throws JOSEException, ParseException {
+        // Input string is configured in build.gradle
+        return signHmacJwt(claimsSet, "test-HMAC/secret passphrase-value");
+    }
+
+    private SignedJWT signHmacJwt(JWTClaimsSet claimsSet, String hmacPassphrase) throws JOSEException, ParseException {
+        final OctetSequenceKey hmac = JwkValidateUtil.buildHmacKeyFromString(hmacPassphrase);
+        final JWSSigner signer = new MACSigner(hmac);
+        return signJWt(signer, "HS256", claimsSet);
+    }
+
+    // JWT construction
+    private JWTClaimsSet buildJwt(Map<String, Object> claims, Instant issueTime) {
+        final JWTClaimsSet.Builder builder = new JWTClaimsSet.Builder();
+        builder.issuer(randomAlphaOfLengthBetween(4, 24));
+        builder.subject(randomAlphaOfLengthBetween(4, 24));
+        builder.audience(randomList(1, 6, () -> randomAlphaOfLengthBetween(4, 12)));
+        if (randomBoolean()) {
+            builder.jwtID(UUIDs.randomBase64UUID(random()));
+        }
+
+        issueTime = issueTime.truncatedTo(ChronoUnit.SECONDS);
+        builder.issueTime(Date.from(issueTime));
+        builder.expirationTime(Date.from(issueTime.plusSeconds(randomLongBetween(180, 1800))));
+        if (randomBoolean()) {
+            builder.notBeforeTime(Date.from(issueTime.minusSeconds(randomLongBetween(10, 90))));
+        }
+
+        // This may overwrite the aud/sub/iss set above. That is an intended behaviour
+        for (String key : claims.keySet()) {
+            builder.claim(key, claims.get(key));
+        }
+        return builder.build();
```

Comment:
Missing optional "typ" header and "auth_time" claim. Both are validated in JwtValidateUtil.validate().

JwtTestCase.buildUnsignedJwt() has code for building JWT header and claims set.
JwtRealmTestCase.randomJwt() has code for randomizing claims passed into buildUnsignedJwt().

Reply:
Caveat: I just noticed randomJwt() does not pass a randomize "typ" or null header into buildUnsignedJwt(). I missed that in a previous PR. JwtValidateUtil.JWT_HEADER_TYPE_VERIFIER is used to check for "typ" header value of "jwt" or null.